src/jit/flowgraph.cpp

```diff
@@ -1621,15 +1694,152 @@ void Compiler::fgReplacePred(BasicBlock* block, BasicBlock* oldPred, BasicBlock*
     noway_assert(newPred != nullptr);
     assert(!fgCheapPredsValid);
 
-    flowList* pred;
+    // Make sure the list is sorted before manipulating.
+    bool sorted = fgExpensiveSortedPredCheck(block);
 
-    for (pred = block->bbPreds; pred != nullptr; pred = pred->flNext)
+#if DEBUG
+    // Also make sure that our refCount is correct before modifying.
+    // If it is not, keep track of that so that we know replacePreds is
+    // not the problem.
+
+    bool refCountCorrect = true;
+
+    if (JitConfig.JitSlowDebugChecksEnabled() != 0)
     {
-        if (oldPred == pred->flBlock)
+        unsigned refCount = 0;
+
+        // Make sure the ref count to this block is still correct.
+        for (flowList* pred = block->bbPreds; pred; pred = pred->flNext)
         {
-            pred->flBlock = newPred;
+            refCount += pred->flDupCount;
+        }
+
+        if (refCount != block->bbRefs)
+        {
+            refCountCorrect = false;
+        }
+    }
+
+#endif // DEBUG
+
+    flowList** elementBeforeMovingPred;
+    flowList*  movingPred = fgGetPredForBlock(block, oldPred, &elementBeforeMovingPred);
+
+    // If the pred we are trying to replace does not exist.
+    // Bail.
+    //
+    // TODO, investigate why we attempt to replace a pred that does not exist.
+    // and change this to an assert.
+    //
+    // Tracked with https://github.com/dotnet/coreclr/issues/13345
+    if (!movingPred)
+        return;
+
+    // Splice out the node that we are modifying.
+    if (*elementBeforeMovingPred)
+    {
+        *elementBeforeMovingPred = movingPred->flNext;
+    }
```

Comment:
But since the list is singly-linked, you can't look before you. You could only avoid unlink/re-link if the replacement block is `> movingPred->flBlock->bbNum` and `< movingPred->flNext->flBlock->bbNum` (which of course would require a null check on `movingPred->flNext`.)